plugin/pkg/client/auth/oidc/oidc.go

```diff
@@ -75,57 +110,61 @@ func newOIDCAuthProvider(_ string, cfg map[string]string, persister restclient.A
 		return nil, fmt.Errorf("Must provide %s", cfgClientSecret)
 	}
 
-	var certAuthData []byte
-	var err error
-	if cfg[cfgCertificateAuthorityData] != "" {
-		certAuthData, err = base64.StdEncoding.DecodeString(cfg[cfgCertificateAuthorityData])
-		if err != nil {
-			return nil, err
+	client, ok := cache.getClient(issuer, clientID, clientSecret)
+	if !ok {
+		var certAuthData []byte
+		var err error
+		if cfg[cfgCertificateAuthorityData] != "" {
+			certAuthData, err = base64.StdEncoding.DecodeString(cfg[cfgCertificateAuthorityData])
+			if err != nil {
+				return nil, err
+			}
 		}
-	}
-
-	clientConfig := restclient.Config{
-		TLSClientConfig: restclient.TLSClientConfig{
-			CAFile: cfg[cfgCertificateAuthority],
-			CAData: certAuthData,
-		},
-	}
 
-	trans, err := restclient.TransportFor(&clientConfig)
-	if err != nil {
-		return nil, err
-	}
-	hc := &http.Client{Transport: trans}
+		clientConfig := restclient.Config{
+			TLSClientConfig: restclient.TLSClientConfig{
+				CAFile: cfg[cfgCertificateAuthority],
+				CAData: certAuthData,
+			},
+		}
 
-	providerCfg, err := oidc.FetchProviderConfig(hc, issuer)
-	if err != nil {
-		return nil, fmt.Errorf("error fetching provider config: %v", err)
-	}
+		trans, err := restclient.TransportFor(&clientConfig)
+		if err != nil {
+			return nil, err
+		}
+		hc := &http.Client{Transport: trans}
 
-	scopes := strings.Split(cfg[cfgExtraScopes], ",")
-	oidcCfg := oidc.ClientConfig{
-		HTTPClient: hc,
-		Credentials: oidc.ClientCredentials{
-			ID:     clientID,
-			Secret: clientSecret,
-		},
-		ProviderConfig: providerCfg,
-		Scope:          append(scopes, oidc.DefaultScope...),
-	}
+		providerCfg, err := oidc.FetchProviderConfig(hc, issuer)
+		if err != nil {
+			return nil, fmt.Errorf("error fetching provider config: %v", err)
+		}
 
-	client, err := oidc.NewClient(oidcCfg)
-	if err != nil {
-		return nil, fmt.Errorf("error creating OIDC Client: %v", err)
+		scopes := strings.Split(cfg[cfgExtraScopes], ",")
+		oidcCfg := oidc.ClientConfig{
+			HTTPClient: hc,
+			Credentials: oidc.ClientCredentials{
+				ID:     clientID,
+				Secret: clientSecret,
+			},
+			ProviderConfig: providerCfg,
+			Scope:          append(scopes, oidc.DefaultScope...),
+		}
+		client, err = oidc.NewClient(oidcCfg)
+		if err != nil {
+			return nil, fmt.Errorf("error creating OIDC Client: %v", err)
+		}
+		cache.setClient(issuer, clientID, clientSecret, client)
```

Comment:
still have the potential to construct this multiple times if newOIDCAuthProvider is called from multiple threads... are we ok with that?